From e35042a506bad59ed7902a04e9155de608f23c54 Mon Sep 17 00:00:00 2001
From: Michael-CY Lee <michael-cy.lee@mediatek.com>
Date: Wed, 15 Jul 2026 15:53:42 +0800
Subject: [PATCH] mtk: hostapd: prevent wpas_ml_connect_pref() from returning
 NULL

wpa_bss_get() and wpa_bss_get_bssid() can return NULL when the
preferred link's BSS is not present in the scan results.
Introduce target_bss for the lookup result and fall back to the
original bss when the preferred link cannot be found.

Signed-off-by: Michael-CY Lee <michael-cy.lee@mediatek.com>
---
 wpa_supplicant/sme.c | 17 +++++++++++++----
 1 file changed, 13 insertions(+), 4 deletions(-)

diff --git a/wpa_supplicant/sme.c b/wpa_supplicant/sme.c
index 0afae3b9c..d341671b1 100644
--- a/wpa_supplicant/sme.c
+++ b/wpa_supplicant/sme.c
@@ -415,6 +415,7 @@ static struct wpa_bss * wpas_ml_connect_pref(struct wpa_supplicant *wpa_s,
 					     struct wpa_bss *bss,
 					     struct wpa_ssid *ssid)
 {
+	struct wpa_bss *target_bss;
 	unsigned int low, high, i;
 
 	wpa_printf(MSG_DEBUG,
@@ -483,13 +484,21 @@ found:
 
 	/* Get the BSS entry and do the switch */
 	if (ssid && ssid->ssid_len)
-		bss = wpa_bss_get(wpa_s, wpa_s->links[i].bssid, ssid->ssid,
-				  ssid->ssid_len);
+		target_bss = wpa_bss_get(wpa_s, wpa_s->links[i].bssid,
+					 ssid->ssid, ssid->ssid_len);
 	else
-		bss = wpa_bss_get_bssid(wpa_s, wpa_s->links[i].bssid);
+		target_bss = wpa_bss_get_bssid(wpa_s, wpa_s->links[i].bssid);
+
+	if (!target_bss) {
+		wpa_printf(MSG_DEBUG,
+			   "MLD: No match for connect/band pref BSS in the"
+			   " scan results");
+		return bss;
+	}
+
 	wpa_s->mlo_assoc_link_id = i;
 
-	return bss;
+	return target_bss;
 }
 #endif /* CONFIG_TESTING_OPTIONS */
 
-- 
2.45.2

